python/kvikio/_lib/libnvcomp.pyx

```diff
@@ -57,305 +63,184 @@ class pyNvcompType_t(Enum):
     pyNVCOMP_TYPE_BITS = nvcompType_t.NVCOMP_TYPE_BITS
 
 
-# _Cascaded Compressor / Decompressor
-cdef class _CascadedCompressor:
-    cdef __CascadedCompressor* c
+cdef class _nvcompManager:
+    # Temporary storage for factory allocated manager to prevent cleanup
+    cdef shared_ptr[nvcompManagerBase] _mgr
+    cdef nvcompManagerBase* _impl
+    cdef shared_ptr[CompressionConfig] _compression_config
+    cdef shared_ptr[DecompressionConfig] _decompression_config
 
-    def __cinit__(self, nvcompType_t t, int num_RLEs, int num_deltas, bool use_bp):
-        self.c = new __CascadedCompressor(
-            t,
-            num_RLEs,
-            num_deltas,
-            use_bp
+    def configure_compression(self, decomp_buffer_size):
+        cdef shared_ptr[CompressionConfig] partial = make_shared[CompressionConfig](
+            self._impl.configure_compression(decomp_buffer_size)
         )
-
-    def __dealloc__(self):
-        del self.c
-
-    def configure(self, in_bytes, temp_bytes, out_bytes):
-        self.c.configure(
-            in_bytes,
-            <size_t*>Array(temp_bytes).ptr,
-            <size_t*>Array(out_bytes).ptr
+        self._compression_config = make_shared[CompressionConfig](
+            (move(partial.get()[0]))
+        )
+        return {
+            "uncompressed_buffer_size": self._compression_config.get()[
+                0
+            ].uncompressed_buffer_size,
+            "max_compressed_buffer_size": self._compression_config.get()[
+                0
+            ].max_compressed_buffer_size,
+            "num_chunks": self._compression_config.get()[0].num_chunks
+        }
+
+    def compress(self, decomp_buffer, comp_buffer):
+        self._impl.compress(
+            <const uint8_t*><uintptr_t>decomp_buffer.data.ptr,
+            <uint8_t*><uintptr_t>comp_buffer.data.ptr,
+            <CompressionConfig&>self._compression_config.get()[0]
+        )
+        size = self._impl.get_compressed_output_size(
+            <uint8_t*><uintptr_t>comp_buffer.data.ptr
         )
+        return size
```

Comment:
Because the buffer has already been allocated, and prior to completing compression it has an unknown size. A possible refactor would be for the library to create the compression storage buffer and then return it, sliced to the correct size, when compression is finished.